### PR TITLE
Type definitions for 'Symbols as WeakMap keys' (#76)

### LINKED ISSUE #52534: Type definitions for 'Symbols as WeakMap keys'
# lib Update Request for 'Symbols as WeakMap keys'

## Configuration Check

My compilation *target* is `ESNext` and my *lib* is `the default`.

## Missing / Incorrect Definition

[Symbols as WeakMap keys](https://github.com/tc39/proposal-symbols-as-weakmap-keys) went to stage 4 at TC39 yesterday.
The types for `WeakMap`, `WeakSet`, `WeakRef` and `FinalizationRegistry` can be updated to reflect that `symbol` is now also allowed alongside the existing `object` constraint.

## Sample Code

```ts
new WeakMap<symbol, any>();
new WeakSet<symbol>();
new WeakRef(Symbol());
new FinalizationRegistry(() => {}).register(Symbol());
```

## Documentation Link

https://github.com/tc39/proposal-symbols-as-weakmap-keys

### PR DESCRIPTION
**Pull request by leoelm**
Fixes #52534.

Adding symbol type support to `WeakMap`, `WeakSet`, `WeakRef` and `FinalizationRegistry`.
The above has been achieved by retrospectively introducing an interface, `WeakTypes` used for the definition of compatible types for the above mentioned structures.

### CODE REVIEW

**Review by rbuckton — APPROVED** [review]
I can re-review and merge once conflicts are resolved.

**Comment by leoelm** [comment]
Rebase is now done and all seems to be looking good still.

**Review by rbuckton — APPROVED** [review]

File: src/lib/es2015.collection.d.ts [hunk 1]
```diff
@@ -58,14 +58,14 @@ interface WeakMap<K extends object, V> {
...
-     * @param key Must be an object.
+     * @param key Must be an object or symbol.
      */
     set(key: K, value: V): this;
 }
 
 interface WeakMapConstructor {
-    new <K extends object = object, V = any>(entries?: readonly (readonly [K, V])[] | null): WeakMap<K, V>;
-    readonly prototype: WeakMap<object, any>;
+    new <K extends WeakKey = WeakKey, V = any>(entries?: readonly [K, V][] | null): WeakMap<K, V>;
```
Comment (graphemecluster) [inline src/lib/es2015.collection.d.ts:67]: Regression from #50449.

Comment (rbuckton) [reply src/lib/es2015.collection.d.ts:67]: Ah, thanks for catching this. I missed it in the review.

Comment (rbuckton) [reply src/lib/es2015.collection.d.ts:67]: https://github.com/microsoft/TypeScript/pull/54592 will fix this

**Pull request merged**